### ISSUE TITLE
Question about Radio Frequency & CAT Control

### ISSUE THREAD

**Issue by Nammurg** [haiduc32/MagLoopTuner#1]
In MagLoopPredict a comment states that this is configured for Elecraft radios, however, I do not understand how the Arduino is able to communicate with the radio over CAT to get the current frequency?

Do I have to connect the CAT interface of the radio to specific pins on the Arduino board or can the Arduino get the data from the radio by being connected to the PC that controls the radio via USB?

**Comment by haiduc32** [comment]
https://github.com/haiduc32/MagLoopTuner/blob/main/MagLoopPredict/MagLoopPredict.ino#L328 that function reads the frequency using CAT.

https://github.com/haiduc32/MagLoopTuner/blob/main/MagLoopPredict/MagLoopPredict.ino#L66 that is the configuration for the serial for Elecraft radios. (I don't remember what is the pinout mapping, will have to document that).

With some modification you could also make it communicate with a pc to get the data, or over network.

The project is not really up to date, I believe I have newere code on my laptop. I plan on getting back and finishing this project in the next months, so just make sure to watch or better yet follow it on https://hackaday.io/project/183721-magloop-auto-tuner

**Comment by Nammurg** [comment]
Thank you for the swift reply and the helpful links! 

I like your project very much as I use a 2E0ERO mag loop this project will just transform the antenna!

Looking forward for the next updates! 

Do you know if the commands on the first link are looking for the Frequency using CI-V addresses?

**Comment by Nammurg** [comment]
In regards of the pin out, from your video, I can see that you have connected 3 out of the 4 connections from the ACC socket of the radio but can't tell where they go on the Arduino board...

**Comment by Nammurg** [comment]
I have done some snooping around and I found that in lines 329 and 330, the serial connection calls for "FA" (for Elecraft radios that is VFO A Frequency) so in theory, if this is changed to "03" which is the current frequency for iCOM radios as well as the baud rate changed to 19200, it should get the correct thing...

Also, the pinout for the ACC -> Arduino connection is Pin 4 for TX and pin 5 for RX

**Comment by haiduc32** [comment]
https://github.com/haiduc32/MagLoopTuner/blob/main/MagLoopPredict/MagLoopPredict.ino#L66 that is the line with the pinout for arduino. I had to use a software serial, as I was using the other one for debug. Third line is ground.

**Important**! for Elecraft I had to use inverted signal. If it doesn't work for your icom, try setting it to false, so it's non-inverted.

**Comment by Nammurg** [comment]
Hello again, I was checking the iCOM manual and it mentions the CIV code for reading the SWR from the radio. 

Would this be achievable with your code so you won't need to do a full frequency calibration? 
This way it will adjust the mag loop capacitor on-the-fly based on the SWR as SWR can change with various conditions like weather and obstacles especially when used indoors...

Also, that will give you an almost 1:1 SWR as it will automatically try to get as close to 1:1 as possible.

Let me know your thoughts please.

**Comment by haiduc32** [comment]
That is interesting. Did you try it? does your radio actually support it? how fast is it?

That would be great if it works. But I'm a bit skeptical. The problems I see:
- you need to sweep the whole range to find the smallest swr. Or at least until it starts going under some threshold, and then put some algorithm in place to tune in the best spot
- step sizes and speed of checking the SWR - if steps are too big, you might jump over the dip and keep going, if swr reading speed is too big, you'll have time to brew a cup of tea while waiting for it to tune
- maybe it can be used together with pre-calibration, so it get's to a close SWR and final adjustment is done with reading the actual SWR

But hey, this hobby is all about experimenting. Maybe you're onto something interesting ;)

Just occured to me, even Elecraft KX2 has a CAT function for reading SWR, but you need to transmit to be able to read SWR. So that's not an option. Check exactly what that CAT command does. 

Paste a link to the command, I'll also take a look.

**Comment by Nammurg** [comment]
Thank you for the swift reply! 

In regards of the frequency range, I use a 2E0ERO Mag Loop (https://www.qrz.com/db/2E0ERO) and I get a bandwidth between 3-10 KHz depending on the band I use (40-30-20-17-15). 

We might be able to put a code in to say if SWR more than 1.50 move motor to the left or to the right and find the lowest SWR point.

This might be an extra feature for when you do the frequency sweep to calibrate / tune the motor.

I have a Xiegu X6100 (which is currently sent for repair) and it uses the same CAT/CIV as iCOM.
[Xiegu CAT Manual](https://radioddity.s3.amazonaws.com/Radioddity%20-%20Xiegu%20G90,%20CAT%20and%20Digital%20modes%20V1.0_20210623.pdf)

Also the radio has an option where it can be put into SWR sweep mode and graph the SWR on the screen.

In a nutshell these are the commands:

0x03 - Get active VFO frequency
0x15 0x12 - Get SWR-Meter
0x1C - Get PTT switch
0x1C 0x00 0x00 - Release PTT
0x1C 0x00 0x01 - Press PTT

As soon as I get my radio back from China, I will start testing these and post an update.

**Comment by Nammurg** [comment]
Also, we might be able to play clever here and on the calibration, we calibrate based on the band plan so you will get ranges for the frequencies and their use so when you do an SWR scan, you will be able to narrow down to a specific range which will speed up the operation.

I mainly use the Digital FT8/FT4 portion of the bands so calibrating for these should be quite fast and easy (...he said... :p )

**Comment by haiduc32** [comment]
Indeed 6100 has a SWR meter, it's not that good from what I've heard. Well.. I'm comparing to a RigExpert now.. So yes, that might work.

I did have a check in the code and it's all up to date.

My antenna right now is in... pieces.. and I can't find an arduino board to set it all back up, but maybe I can just find the energy need to get it all back up and finishing the work on at least first fully working version. And documenting it all. Fingers crossed.

I don't remember exactly what was my issue why I left it for so long, but it might be because I could never get a 100% homing. But in my book if I can get it under 2 SWR, I'll call it acceptable.

**Comment by Nammurg** [comment]
I have no extra equipment that I can verify the SWR on any radio so I would like to know what your findings are!

I have couple of spare Arduino Mega boards so if you are located in the UK, I can send you one if you want...

I really love your project as it will help me immensely automating  the mag loop tuning! 

Thank you for everything!

**Comment by haiduc32** [comment]
I should have about 3 arduino unos, found one but that is put aside for another project. so it's a matter of sorting out lots of stuff to find the other 2. Would be fun to try to make contact with the setup, as I get UK contacts pretty often from Romania on the 20m band.

**Comment by Nammurg** [comment]
In regards of the motor, do I have to put a physical stop or should I leave it spinning freely?

I mean how does it know which way is home, the 0 starting point?

**Comment by haiduc32** [comment]
The stepper motor should be turning the capacitor for tuning. The capacitor is supposed to have a physical stop. Mine has about 2.5 full rotations. The stepper motor in combination with the driver detects when it reaches the stop and stalls. (that bit of code is not what I wanted, but it works...)

Current code assumes that rotating counter-clock-wise will tune lower, and clock-wise higher. It will home to the left at most 3 full turns (from my memory).

**Comment by Nammurg** [comment]
Oh that's interesting, the capacitor on my antenna does not have a stop at all!

Need to incorporate one and try again!

**Comment by haiduc32** [comment]
before doing un-reversible modifications, consider if there is any other way you can adapt your setup to have a home detection, maybe something that can actuate a switch when at low or high position.

**Comment by haiduc32** [comment]
I've been thinking. If you're idea with reading SWR from the radio works, then you don't actually need to "home" the servo.
The code needs some significant change, but you could just "seek" a best calibration point when initially starting - might take some time - and then just tune up or down as you scroll through frequencies. Probably put in a limit on number of revolution to "seek" so it doesn't rotate infinitely if a good enough match isnt' found.

**Comment by Nammurg** [comment]
Yes, good point!
I managed to do a mechanical stop by using couple of screws. 
It works well. 
I am wondering if there is any way you can move the motor via the command line?
Like moveTo position 500 etc...

**Comment by haiduc32** [comment]
Check this sketch, which I initially used for testing and calibration: https://github.com/haiduc32/MagLoopTuner/blob/main/MagLoopCal/MagLoopCal.ino
It should be working.. I think. This should be enough for you to debug. Has some terminal commands too.

Introducing terminal commands in the main sketch is possible but not that easy. I don't want to go down that road.

**Comment by Nammurg** [comment]
Thank you, I'll have a look, very helpful!

My antenna needs to be very fine tuned like in millimetres as any major moves, the frequency changes tremendously.

To give you an idea, I had to change the step from 50 down to 2 as with one click on 50 the frequency changed by 500KHz!
So having a go-to function in the command line I can speed up the calibration as I can point the motor to the rough frequency start and then fine tune from there.

Is there any way to delete the EEPROM calibration data?

**Comment by haiduc32** [comment]
Short answer: I don't remember. Simplest solution I can think of without digging into code is to change the MAGIC number: https://github.com/haiduc32/MagLoopTuner/blob/main/MagLoopPredict/MagLoopPredict.ino#L52
But I think if you enter calibration again, it will get overridden anyway.

Good news is I finally figured the problem with my antenna, after a few very frustrating days. I melted the inner core of the small loop :|. That's what happens when you push too many watts through a mag loop. Did a temporary fix, so will be able to slowly restart the project. Bad news is I'm in very short supply of free time, so updates will be coming slow.

**Comment by Nammurg** [comment]
No worries bud, I have found that changing the number 50 in line 265 to something smaller did the trick.
https://github.com/haiduc32/MagLoopTuner/blob/1680994ad4cc5edf03a940896c18fb58eff57ace/MagLoopPredict/MagLoopPredict.ino#L265

Don't worry out the time line, we can contribute and test as much as we can. ;)

to give you an idea for my magloop's capacitor fine tuning, I had the whole 20m band between motor steps 780 and 800!
Started at 13.800Mhz all the way to 14.500MHz respectively.

One problem that I have noticed in the testing is that the motor stop I created (the one with the screw in the shaft) it is not always accurate and the 0 start point is not always the same resulting in the steps being off by 30-50.
This coupled with the extremely fine tune needed by the specific magloop, makes it impossible to calibrate accurately and have repetitive results.

**Comment by haiduc32** [comment]
Stepper motors have 200 steps in a revolution. But then there are micro-steps. From memory, you can have up to 256 micro steps in a step - depends on your stepper driver.  If you use the one specified in the sketch, it does support up to 256 micro steps (https://www.trinamic.com/products/integrated-circuits/details/tmc2130/). 

The current setup uses 16 micro steps per step. That means you can go another 16 micro-steps for each steps as it is now. That's where it is configured: https://github.com/haiduc32/MagLoopTuner/blob/main/MagLoopPredict/MagLoopPredict.ino#L137

But beware, stuff will work differently when you change that. Detecting stall might not work anymore or would need finding the magic number that will work.

Further ideas, you could find a worm gear, or a reduction gear setup. With SWR feedback, that would work. Just with calibration data points that creates play that increases uncertainty.
